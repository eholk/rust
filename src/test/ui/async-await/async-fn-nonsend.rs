// edition:2018
// compile-flags: --crate-type lib

use std::{fmt::Debug, rc::Rc};

fn non_send() -> impl Debug {
    Rc::new(())
}

async fn fut() {}

async fn non_send_temporary_in_match() {
    // We could theoretically make this work as well (produce a `Send` future)
    // for scrutinees / temporaries that can or will
    // be dropped prior to the match body
    // (e.g. `Copy` types).
    match Some(non_send()) {
        Some(_) => fut().await,
        None => {}
    }
}

fn assert_send(_: impl Send) {}

pub fn pass_assert() {
    assert_send(non_send_temporary_in_match());
    //~^ ERROR future cannot be sent between threads safely
    //~^^ ERROR future cannot be sent between threads safely
}
