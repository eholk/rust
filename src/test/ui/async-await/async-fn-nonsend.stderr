error: future cannot be sent between threads safely
  --> $DIR/async-fn-nonsend.rs:50:5
   |
LL |     assert_send(non_send_temporary_in_match());
   |     ^^^^^^^^^^^ future returned by `non_send_temporary_in_match` is not `Send`
   |
   = help: within `impl Future`, the trait `Send` is not implemented for `Rc<()>`
note: future is not `Send` as this value is used across an await
  --> $DIR/async-fn-nonsend.rs:33:20
   |
LL |     match Some(non_send()) {
   |           ---------------- has type `Option<impl Debug>` which is not `Send`
LL |         Some(_) => fut().await,
   |                    ^^^^^^^^^^^ await occurs here, with `Some(non_send())` maybe used later
...
LL | }
   | - `Some(non_send())` is later dropped here
note: required by a bound in `assert_send`
  --> $DIR/async-fn-nonsend.rs:46:24
   |
LL | fn assert_send(_: impl Send) {}
   |                        ^^^^ required by this bound in `assert_send`

error: future cannot be sent between threads safely
  --> $DIR/async-fn-nonsend.rs:52:5
   |
LL |     assert_send(non_sync_with_method_call());
   |     ^^^^^^^^^^^ future returned by `non_sync_with_method_call` is not `Send`
   |
   = help: the trait `Send` is not implemented for `dyn std::fmt::Write`
note: future is not `Send` as this value is used across an await
  --> $DIR/async-fn-nonsend.rs:42:9
   |
LL |     let f: &mut std::fmt::Formatter = panic!();
   |         - has type `&mut Formatter<'_>` which is not `Send`
LL |     if non_sync().fmt(f).unwrap() == () {
LL |         fut().await;
   |         ^^^^^^^^^^^ await occurs here, with `f` maybe used later
LL |     }
LL | }
   | - `f` is later dropped here
note: required by a bound in `assert_send`
  --> $DIR/async-fn-nonsend.rs:46:24
   |
LL | fn assert_send(_: impl Send) {}
   |                        ^^^^ required by this bound in `assert_send`

error: aborting due to 2 previous errors

